{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "5d-QZmJTjRF-",
        "outputId": "214c3cf1-6d2b-494e-b9ad-d132166d6702"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Collecting Bio\n",
            "  Obtaining dependency information for Bio from https://files.pythonhosted.org/packages/34/bd/5077f1d9f5a9508edae77a9647228a7304457d22469b2fe88ec7c990f10f/bio-1.6.0-py3-none-any.whl.metadata\n",
            "  Downloading bio-1.6.0-py3-none-any.whl.metadata (5.1 kB)\n",
            "Collecting biopython>=1.80 (from Bio)\n",
            "  Downloading biopython-1.81-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m3.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m0m\n",
            "\u001b[?25hRequirement already satisfied: requests in /home/lab/anaconda3/lib/python3.11/site-packages (from Bio) (2.31.0)\n",
            "Requirement already satisfied: tqdm in /home/lab/anaconda3/lib/python3.11/site-packages (from Bio) (4.65.0)\n",
            "Collecting mygene (from Bio)\n",
            "  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
            "Requirement already satisfied: pandas in /home/lab/anaconda3/lib/python3.11/site-packages (from Bio) (2.0.3)\n",
            "Collecting pooch (from Bio)\n",
            "  Obtaining dependency information for pooch from https://files.pythonhosted.org/packages/1a/a5/5174dac3957ac412e80a00f30b6507031fcab7000afc9ea0ac413bddcff2/pooch-1.8.0-py3-none-any.whl.metadata\n",
            "  Downloading pooch-1.8.0-py3-none-any.whl.metadata (9.9 kB)\n",
            "Collecting gprofiler-official (from Bio)\n",
            "  Downloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n",
            "Requirement already satisfied: numpy in /home/lab/anaconda3/lib/python3.11/site-packages (from biopython>=1.80->Bio) (1.24.3)\n",
            "Collecting biothings-client>=0.2.6 (from mygene->Bio)\n",
            "  Obtaining dependency information for biothings-client>=0.2.6 from https://files.pythonhosted.org/packages/c2/21/6c4bdb8ba8d2cdeb5ac3a6460ab1cbd841e46cd851d6b00028b327c5deb3/biothings_client-0.3.1-py2.py3-none-any.whl.metadata\n",
            "  Downloading biothings_client-0.3.1-py2.py3-none-any.whl.metadata (9.8 kB)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /home/lab/anaconda3/lib/python3.11/site-packages (from pandas->Bio) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /home/lab/anaconda3/lib/python3.11/site-packages (from pandas->Bio) (2023.3.post1)\n",
            "Requirement already satisfied: tzdata>=2022.1 in /home/lab/anaconda3/lib/python3.11/site-packages (from pandas->Bio) (2023.3)\n",
            "Requirement already satisfied: platformdirs>=2.5.0 in /home/lab/anaconda3/lib/python3.11/site-packages (from pooch->Bio) (3.10.0)\n",
            "Requirement already satisfied: packaging>=20.0 in /home/lab/anaconda3/lib/python3.11/site-packages (from pooch->Bio) (23.1)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /home/lab/anaconda3/lib/python3.11/site-packages (from requests->Bio) (2.0.4)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /home/lab/anaconda3/lib/python3.11/site-packages (from requests->Bio) (3.4)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /home/lab/anaconda3/lib/python3.11/site-packages (from requests->Bio) (1.26.16)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /home/lab/anaconda3/lib/python3.11/site-packages (from requests->Bio) (2023.7.22)\n",
            "Requirement already satisfied: six>=1.5 in /home/lab/anaconda3/lib/python3.11/site-packages (from python-dateutil>=2.8.2->pandas->Bio) (1.16.0)\n",
            "Downloading bio-1.6.0-py3-none-any.whl (279 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m279.4/279.4 kB\u001b[0m \u001b[31m48.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pooch-1.8.0-py3-none-any.whl (62 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m62.7/62.7 kB\u001b[0m \u001b[31m19.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n",
            "Installing collected packages: biopython, pooch, gprofiler-official, biothings-client, mygene, Bio\n",
            "Successfully installed Bio-1.6.0 biopython-1.81 biothings-client-0.3.1 gprofiler-official-1.0.0 mygene-3.2.2 pooch-1.8.0\n"
          ]
        }
      ],
      "source": [
        "!pip install Bio"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ISlylvRMhVIl"
      },
      "outputs": [],
      "source": [
        "from Bio import SeqIO\n",
        "import os\n",
        "import pandas as pd\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "BSJaVY-BhVIo",
        "outputId": "737b627d-e20a-486f-b695-3aa3f34bdabb"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Error en el archivo LT856801.gbk: Problem with 'source' feature:\n",
            "1..426\n",
            "/PCR_primers=\"fwd_name: 341F, fwd_seq:\n",
            "nnnnnnnnntcctacgggnggcwgcag, rev_name: 785R, rev_seq:\n",
            "nnnnnnnnnntgactachvgggtatctaakcc\"\n",
            "/db_xref=\"taxon:158751\"\n",
            "/environmental_sample\n",
            "/isolate=\"OTU119\"\n",
            "/\n",
            "isolation_source=\"Anaerobic_CO2_and_current_fed_bioreacto\n",
            "r\"\n",
            "/mol_type=\"genomic DNA\"\n",
            "/organism=\"uncultured Acidovorax sp.\"\n",
            "Error en el archivo LT859231.gbk: Problem with 'source' feature:\n",
            "1..426\n",
            "/PCR_primers=\"fwd_name: 341F, fwd_seq:\n",
            "nnnnnnnnntcctacgggnggcwgcag, rev_name: 785R, rev_seq:\n",
            "nnnnnnnnnntgactachvgggtatctaakcc\"\n",
            "/db_xref=\"taxon:158751\"\n",
            "/environmental_sample\n",
            "/isolate=\"OTU2549\"\n",
            "/\n",
            "isolation_source=\"Anaerobic_CO2_and_current_fed_bioreacto\n",
            "r\"\n",
            "/mol_type=\"genomic DNA\"\n",
            "/organism=\"uncultured Acidovorax sp.\"\n",
            "Error en el archivo LT857898.gbk: Problem with 'source' feature:\n",
            "1..423\n",
            "/PCR_primers=\"fwd_name: 341F, fwd_seq:\n",
            "nnnnnnnnntcctacgggnggcwgcag, rev_name: 785R, rev_seq:\n",
            "nnnnnnnnnntgactachvgggtatctaakcc\"\n",
            "/db_xref=\"taxon:165188\"\n",
            "/environmental_sample\n",
            "/isolate=\"OTU1216\"\n",
            "/\n",
            "isolation_source=\"Anaerobic_CO2_and_current_fed_bioreacto\n",
            "r\"\n",
            "/mol_type=\"genomic DNA\"\n",
            "/organism=\"uncultured Megasphaera sp.\"\n",
            "Error en el archivo LT857056.gbk: Problem with 'source' feature:\n",
            "1..420\n",
            "/PCR_primers=\"fwd_name: 341F, fwd_seq:\n",
            "nnnnnnnnntcctacgggnggcwgcag, rev_name: 785R, rev_seq:\n",
            "nnnnnnnnnntgactachvgggtatctaakcc\"\n",
            "/db_xref=\"taxon:171292\"\n",
            "/environmental_sample\n",
            "/isolate=\"OTU374\"\n",
            "/\n",
            "isolation_source=\"Anaerobic_CO2_and_current_fed_bioreacto\n",
            "r\"\n",
            "/mol_type=\"genomic DNA\"\n",
            "/organism=\"uncultured Mycobacterium sp.\"\n",
            "Error en el archivo LT856882.gbk: Problem with 'source' feature:\n",
            "1..406\n",
            "/PCR_primers=\"fwd_name: 341F, fwd_seq:\n",
            "nnnnnnnnntcctacgggnggcwgcag, rev_name: 785R, rev_seq:\n",
            "nnnnnnnnnntgactachvgggtatctaakcc\"\n",
            "/db_xref=\"taxon:171292\"\n",
            "/environmental_sample\n",
            "/isolate=\"OTU200\"\n",
            "/\n",
            "isolation_source=\"Anaerobic_CO2_and_current_fed_bioreacto\n",
            "r\"\n",
            "/mol_type=\"genomic DNA\"\n",
            "/organism=\"uncultured Mycobacterium sp.\"\n",
            "Error en el archivo LT859290.gbk: Problem with 'source' feature:\n",
            "1..426\n",
            "/PCR_primers=\"fwd_name: 341F, fwd_seq:\n",
            "nnnnnnnnntcctacgggnggcwgcag, rev_name: 785R, rev_seq:\n",
            "nnnnnnnnnntgactachvgggtatctaakcc\"\n",
            "/db_xref=\"taxon:158751\"\n",
            "/environmental_sample\n",
            "/isolate=\"OTU2608\"\n",
            "/\n",
            "isolation_source=\"Anaerobic_CO2_and_current_fed_bioreacto\n",
            "r\"\n",
            "/mol_type=\"genomic DNA\"\n",
            "/organism=\"uncultured Acidovorax sp.\"\n",
            "Error en el archivo LT857903.gbk: Problem with 'source' feature:\n",
            "1..426\n",
            "/PCR_primers=\"fwd_name: 341F, fwd_seq:\n",
            "nnnnnnnnntcctacgggnggcwgcag, rev_name: 785R, rev_seq:\n",
            "nnnnnnnnnntgactachvgggtatctaakcc\"\n",
            "/db_xref=\"taxon:159268\"\n",
            "/environmental_sample\n",
            "/isolate=\"OTU1221\"\n",
            "/\n",
            "isolation_source=\"Anaerobic_CO2_and_current_fed_bioreacto\n",
            "r\"\n",
            "/mol_type=\"genomic DNA\"\n",
            "/organism=\"uncultured Veillonella sp.\"\n",
            "Error en el archivo LT857902.gbk: Problem with 'source' feature:\n",
            "1..426\n",
            "/PCR_primers=\"fwd_name: 341F, fwd_seq:\n",
            "nnnnnnnnntcctacgggnggcwgcag, rev_name: 785R, rev_seq:\n",
            "nnnnnnnnnntgactachvgggtatctaakcc\"\n",
            "/db_xref=\"taxon:165188\"\n",
            "/environmental_sample\n",
            "/isolate=\"OTU1220\"\n",
            "/\n",
            "isolation_source=\"Anaerobic_CO2_and_current_fed_bioreacto\n",
            "r\"\n",
            "/mol_type=\"genomic DNA\"\n",
            "/organism=\"uncultured Megasphaera sp.\"\n",
            "Error en el archivo LT857901.gbk: Problem with 'source' feature:\n",
            "1..426\n",
            "/PCR_primers=\"fwd_name: 341F, fwd_seq:\n",
            "nnnnnnnnntcctacgggnggcwgcag, rev_name: 785R, rev_seq:\n",
            "nnnnnnnnnntgactachvgggtatctaakcc\"\n",
            "/db_xref=\"taxon:165188\"\n",
            "/environmental_sample\n",
            "/isolate=\"OTU1219\"\n",
            "/\n",
            "isolation_source=\"Anaerobic_CO2_and_current_fed_bioreacto\n",
            "r\"\n",
            "/mol_type=\"genomic DNA\"\n",
            "/organism=\"uncultured Megasphaera sp.\"\n",
            "Error en el archivo LT856739.gbk: Problem with 'source' feature:\n",
            "1..426\n",
            "/PCR_primers=\"fwd_name: 341F, fwd_seq:\n",
            "nnnnnnnnntcctacgggnggcwgcag, rev_name: 785R, rev_seq:\n",
            "nnnnnnnnnntgactachvgggtatctaakcc\"\n",
            "/db_xref=\"taxon:158751\"\n",
            "/environmental_sample\n",
            "/isolate=\"OTU57\"\n",
            "/\n",
            "isolation_source=\"Anaerobic_CO2_and_current_fed_bioreacto\n",
            "r\"\n",
            "/mol_type=\"genomic DNA\"\n",
            "/organism=\"uncultured Acidovorax sp.\"\n",
            "Error en el archivo LT857900.gbk: Problem with 'source' feature:\n",
            "1..426\n",
            "/PCR_primers=\"fwd_name: 341F, fwd_seq:\n",
            "nnnnnnnnntcctacgggnggcwgcag, rev_name: 785R, rev_seq:\n",
            "nnnnnnnnnntgactachvgggtatctaakcc\"\n",
            "/db_xref=\"taxon:165188\"\n",
            "/environmental_sample\n",
            "/isolate=\"OTU1218\"\n",
            "/\n",
            "isolation_source=\"Anaerobic_CO2_and_current_fed_bioreacto\n",
            "r\"\n",
            "/mol_type=\"genomic DNA\"\n",
            "/organism=\"uncultured Megasphaera sp.\"\n",
            "        Accession                                        Description  \\\n",
            "0      KJ210353.1  Acidovorax citrulli strain LS-2 16S ribosomal ...   \n",
            "1      HQ097125.1  Uncultured Acidovorax sp. clone F5OHPNU07IKY0F...   \n",
            "2      HQ049153.1  Uncultured Acidovorax sp. clone F5OHPNU07ILFGM...   \n",
            "3      LT696105.1  Uncultured Streptococcus sp. partial 16S rRNA ...   \n",
            "4      LT685978.1  Uncultured Prevotella sp. partial 16S rRNA gen...   \n",
            "...           ...                                                ...   \n",
            "44622  LC317311.1  Streptococcus uberis gene for 16S ribosomal RN...   \n",
            "44623  AY325746.1  Uncultured Neisseria sp. isolate PCR amplicon ...   \n",
            "44624  GU406190.1  Veillonella parvula clone ATC_H73_85 16S ribos...   \n",
            "44625  GU405726.1  Veillonella parvula clone PI054 16S ribosomal ...   \n",
            "44626  KX607284.1  Mycobacterium sp. strain 3249 16S ribosomal RN...   \n",
            "\n",
            "                                                Taxonomy  \\\n",
            "0      [Bacteria, Pseudomonadota, Betaproteobacteria,...   \n",
            "1      [Bacteria, Pseudomonadota, Betaproteobacteria,...   \n",
            "2      [Bacteria, Pseudomonadota, Betaproteobacteria,...   \n",
            "3      [Bacteria, Bacillota, Bacilli, Lactobacillales...   \n",
            "4      [Bacteria, Bacteroidota, Bacteroidia, Bacteroi...   \n",
            "...                                                  ...   \n",
            "44622  [Bacteria, Bacillota, Bacilli, Lactobacillales...   \n",
            "44623  [Bacteria, Pseudomonadota, Betaproteobacteria,...   \n",
            "44624  [Bacteria, Bacillota, Negativicutes, Veillonel...   \n",
            "44625  [Bacteria, Bacillota, Negativicutes, Veillonel...   \n",
            "44626  [Bacteria, Actinomycetota, Actinomycetes, Myco...   \n",
            "\n",
            "                      Specie Strain                      Organism  \\\n",
            "0              Paracidovorax    N/A        Paracidovorax citrulli   \n",
            "1      environmental samples    N/A     uncultured Acidovorax sp.   \n",
            "2      environmental samples    N/A     uncultured Acidovorax sp.   \n",
            "3      environmental samples    N/A  uncultured Streptococcus sp.   \n",
            "4      environmental samples    N/A     uncultured Prevotella sp.   \n",
            "...                      ...    ...                           ...   \n",
            "44622          Streptococcus    N/A          Streptococcus uberis   \n",
            "44623  environmental samples    N/A      uncultured Neisseria sp.   \n",
            "44624            Veillonella    N/A           Veillonella parvula   \n",
            "44625            Veillonella    N/A           Veillonella parvula   \n",
            "44626          Mycobacterium    N/A             Mycobacterium sp.   \n",
            "\n",
            "      Modification Date        Country               Isolation Source  \\\n",
            "0           01-OCT-2014  China: Hainan                            N/A   \n",
            "1           04-JAN-2011            N/A            tailing pound 55 ft   \n",
            "2           04-JAN-2011            N/A            tailing pound 15 ft   \n",
            "3           31-JAN-2017            N/A     human head and neck tissue   \n",
            "4           31-JAN-2017            N/A     human head and neck tissue   \n",
            "...                 ...            ...                            ...   \n",
            "44622       17-APR-2018        Ecuador  mastitic milk of dairy cattle   \n",
            "44623       10-FEB-2004            N/A      human cerebrospinal fluid   \n",
            "44624       22-SEP-2010            N/A                    oral cavity   \n",
            "44625       22-SEP-2010            N/A                    oral cavity   \n",
            "44626       31-JUL-2016         Gambia                            N/A   \n",
            "\n",
            "               Host                                           Sequence  \n",
            "0        watermelon  CTTAACATGCAGTCGAACGGTAACAGGTCTTCGGATGCTGACGAGT...  \n",
            "1               N/A  GGTTTAATTCGATGCAGCGCGAAAACCTTACCACTTGACATGTACG...  \n",
            "2               N/A  GACGGGACCGCACAAGCGGTGGATGATGTGGTTTAATTCGATGCAA...  \n",
            "3               N/A  GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTAGAACGCTGAA...  \n",
            "4               N/A  GATGAACGCTGGCTACAGGCTTAACACATGCAAGTCGAGGGGAAAC...  \n",
            "...             ...                                                ...  \n",
            "44622    Bos taurus  TGCCTAATACATGCAAGTAGAACGCTGAGGACTGGTGCTTGCACTG...  \n",
            "44623           N/A  ACATGCAAGTCGGACGGCAGCACAGAGAAGCTTGCTTCTCGGGTGG...  \n",
            "44624  Homo sapiens  GACGAACGCGGCGGCGTGCTTAACACATGCAAGTCGAACGAAGAGC...  \n",
            "44625  Homo sapiens  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGAAGAG...  \n",
            "44626  Homo sapiens  TGAGTAACACGTGGGTGATCTGCCCTGCACTTTGGGATAAGCCTGG...  \n",
            "\n",
            "[44627 rows x 11 columns]\n"
          ]
        }
      ],
      "source": [
        "from Bio import SeqIO\n",
        "import os\n",
        "import pandas as pd\n",
        "\n",
        "directorio_archivos = 'ALL_Bacterias'\n",
        "\n",
        "# Lista para almacenar los datos\n",
        "datos = []\n",
        "\n",
        "# Iterar sobre cada archivo en el directorio\n",
        "for archivo in os.listdir(directorio_archivos):\n",
        "    if archivo.endswith('.gbk'):\n",
        "        ruta_archivo = os.path.join(directorio_archivos, archivo)\n",
        "\n",
        "        try:\n",
        "            # Leer el archivo GenBank\n",
        "            with open(ruta_archivo, 'r') as file:\n",
        "                record = SeqIO.read(file, 'genbank')\n",
        "\n",
        "                # Extraer información relevante\n",
        "                accession = record.id\n",
        "                definition = record.description\n",
        "                taxonomy = record.annotations.get('taxonomy', ['N/A'])\n",
        "                species = taxonomy[-1] if taxonomy else 'N/A'\n",
        "                strain = record.annotations.get('strain', 'N/A')\n",
        "                organism = record.annotations.get('organism', 'N/A')\n",
        "                modification_date = record.annotations.get('date', 'N/A')\n",
        "                country = record.features[0].qualifiers.get('country', ['N/A'])[0]\n",
        "                isolation_source = record.features[0].qualifiers.get('isolation_source', ['N/A'])[0]\n",
        "                host = record.features[0].qualifiers.get('host', ['N/A'])[0]\n",
        "                sequence = str(record.seq)\n",
        "\n",
        "                # Agregar datos a la lista\n",
        "                datos.append({'Accession': accession,\n",
        "                              'Description': definition,\n",
        "                              'Taxonomy': taxonomy,\n",
        "                              'Specie': species,\n",
        "                              'Strain': strain,\n",
        "                              'Organism': organism,\n",
        "                              'Modification Date': modification_date,\n",
        "                              'Country': country,\n",
        "                              'Isolation Source': isolation_source,\n",
        "                              'Host': host,\n",
        "                              'Sequence': sequence})\n",
        "\n",
        "        except ValueError as e:\n",
        "            print(f\"Error en el archivo {archivo}: {e}\")\n",
        "\n",
        "# Crear un DataFrame con los datos\n",
        "df = pd.DataFrame(datos)\n",
        "\n",
        "# Imprimir el DataFrame\n",
        "print(df)\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "base",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.5"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}